### PR TITLE
Add Utf8->Binary type coercion for comparison

### PR DESCRIPTION
**Pull request by jonahgao**
# Which issue does this PR close?

Partial fix for #7039.

Closes https://github.com/apache/arrow-datafusion/issues/7039#

# Rationale for this change
The following casts should be safe:
- Utf8 -> Binary | LargeBinary
- Utf8 | LargetUtf8 | Binary -> LargeBinary

# What changes are included in this PR?

# Are these changes tested?

Yes
# Are there any user-facing changes?

No

### CODE REVIEW

**Review by alamb — APPROVED** [review]
Thanks @jonahgao ❤️ 

I took the liberty of merging up from main to get a fix for CI failures and pushed a small sql level test in sqllogictests

**Review by jackwener — APPROVED** [review]
Thank you @jonahgao

File: datafusion/expr/src/type_coercion/binary.rs [hunk 1]
```diff
@@ -766,6 +767,23 @@ fn string_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType>
...
+    use arrow::datatypes::DataType::*;
+    match (lhs_type, rhs_type) {
+        (Binary, Binary) | (Utf8, Binary) | (Binary, Utf8) => Some(Binary),
+        (LargeBinary, LargeBinary)
+        | (Binary, LargeBinary)
+        | (LargeBinary, Binary)
+        | (Utf8, LargeBinary)
+        | (LargeBinary, Utf8)
+        | (LargeUtf8, LargeBinary)
+        | (LargeBinary, LargeUtf8) => Some(LargeBinary),
```
Comment (jackwener) [inline datafusion/expr/src/type_coercion/binary.rs:782]: we can use 
```rs
(LargeBinary, Utf8 | LargeUtf8 | Binary | LargeBinary) | (Utf8 | LargeUtf8 | Binary | LargeBinary, LargeBinary)
```

Comment (jonahgao) [reply datafusion/expr/src/type_coercion/binary.rs]: Updated. Thank you @jackwener :heart:

**Comment by jonahgao** [comment]
> Thanks @jonahgao ❤️
> 
> I took the liberty of merging up from main to get a fix for CI failures and pushed a small sql level test in sqllogictests

Thank you @alamb  ❤️

**Pull request merged**